### PR TITLE
Return compile-time value instead of instantiating a function

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**

### CODE REVIEW

**Review by czgdp1807 — APPROVED** [review]

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -1647,6 +1647,10 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
     }
 
     void visit_IntrinsicFunction(const ASR::IntrinsicFunction_t& x) {
+        if (x.m_value) {
+            this->visit_expr_wrapper(x.m_value, true);
+            return;
+        }
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_llvm.cpp:1653]: I think this change is causing errors on the CI.

Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_llvm.cpp:1653]: The following fails in `main` itself, but was unnoticed!
```console
$ lfortran integration_tests/intrinsics_20.f90 --fast
```

Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_llvm.cpp:1653]: The following fails in the `main`!
```fortran
program expr2
implicit none

integer :: y
y = 3
print *, cos((y+4)*7.)

end program
```

Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_llvm.cpp:1653]: The issue was related to `inline_function` pass, BindC handling.

@czgdp1807 Would this:[`7854baa` (#2085)](https://github.com/lfortran/lfortran/pull/2085/commits/7854baa549ecb6f684da2c145104581dc1411f63) be a proper fix for this?


<details>
<summary>This PR ASR</summary>


```clojure
(TranslationUnit
    (SymbolTable
        1
        {
            _lcompilers_cos_f32:
                (Function
                    (SymbolTable
                        3
                        {
                            _lcompilers_cos_f32:
                                (Variable
                                    3
                                    _lcompilers_cos_f32
                                    []
                                    ReturnVar
                                    ()
                                    ()
                                    Default
                                    (Real 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            _lfortran_scos:
                                (Function
                                    (SymbolTable
                                        4
                                        {
                                            _lfortran_scos:
                                                (Variable
                                                    4
                                                    _lfortran_scos
                                                    []
                                                    ReturnVar
                                                    ()
                                                    ()
                                                    Default
                                                    (Real 4)
                                                    ()
                                                    BindC
                                                    Public
                                                    Required
                                                    .false.
                                                ),
                                            x:
                                                (Variable
                                                    4
                                                    x
                                                    []
                                                    In
                                                    ()
                                                    ()
                                                    Default
                                                    (Real 4)
                                                    ()
                                                    BindC
                                                    Public
                                                    Required
                                                    .true.
                                                )
                                        })
                                    _lfortran_scos
                                    (FunctionType
                                        [(Real 4)]
                                        (Real 4)
                                        BindC
                                        Interface
                                        "_lfortran_scos"
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        []
                                        .false.
                                    )
                                    []
                                    [(Var 4 x)]
                                    []
                                    (Var 4 _lfortran_scos)
                                    Public
                                    .false.
                                    .false.
                                    ()
                                ),
                            x:
                                (Variable
                                    3
                                    x
                                    []
                                    In
                                    ()
                                    ()
                                    Default
                                    (Real 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    _lcompilers_cos_f32
                    (FunctionType
                        [(Real 4)]
                        (Real 4)
                        Source
                        Implementation
                        ()
                        .false.
                        .false.
                        .false.
                        .false.
                        .false.
                        []
                        .false.
                    )
                    [_lfortran_scos]
                    [(Var 3 x)]
                    [(=
                        (Var 3 _lcompilers_cos_f32)
                        (FunctionCall
                            3 _lfortran_scos
                            3 _lfortran_scos
                            [((Var 3 x))]
                            (Real 4)
                            ()
                            ()
                        )
                        ()
                    )]
                    (Var 3 _lcompilers_cos_f32)
                    Public
                    .false.
                    .false.
                    ()
                ),
            expr2:
                (Program
                    (SymbolTable
                        2
                        {
                            x:
                                (Variable
                                    2
                                    x
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Integer 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    expr2
                    []
                    [(=
                        (Var 2 x)
                        (IntegerConstant 3 (Integer 4))
                        ()
                    )
                    (Print
                        ()
                        [(FunctionCall
                            1 _lcompilers_cos_f32
                            1 _lcompilers_cos_f32
                            [((RealBinOp
                                (Cast
                                    (IntegerBinOp
                                        (Var 2 x)
                                        Add
                                        (IntegerConstant 4 (Integer 4))
                                        (Integer 4)
                                        ()
                                    )
                                    IntegerToReal
                                    (Real 4)
                                    ()
                                )
                                Mul
                                (RealConstant
                                    7.000000
                                    (Real 4)
                                )
                                (Real 4)
                                ()
                            ))]
                            (Real 4)
                            ()
                            ()
                        )]
                        ()
                        ()
                    )]
                )
        })
    []
)
```

</details>

<details>
<summary>Main ASR</summary>

```clojure
(TranslationUnit
    (SymbolTable
        1
        {
            _lcompilers_cos_f32:
                (Function
                    (SymbolTable
                        3
                        {
                            _lcompilers_cos_f32:
                                (Variable
                                    3
                                    _lcompilers_cos_f32
                                    []
                                    ReturnVar
                                    ()
                                    ()
                                    Default
                                    (Real 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            _lfortran_scos:
                                (Function
                                    (SymbolTable
                                        4
                                        {
                                            _lfortran_scos:
                                                (Variable
                                                    4
                                                    _lfortran_scos
                                                    []
                                                    ReturnVar
                                                    ()
                                                    ()
                                                    Default
                                                    (Real 4)
                                                    ()
                                                    BindC
                                                    Public
                                                    Required
                                                    .false.
                                                ),
                                            x:
                                                (Variable
                                                    4
                                                    x
                                                    []
                                                    In
                                                    ()
                                                    ()
                                                    Default
                                                    (Real 4)
                                                    ()
                                                    BindC
                                                    Public
                                                    Required
                                                    .true.
                                                )
                                        })
                                    _lfortran_scos
                                    (FunctionType
                                        [(Real 4)]
                                        (Real 4)
                                        BindC
                                        Interface
                                        "_lfortran_scos"
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        []
                                        .false.
                                    )
                                    []
                                    [(Var 4 x)]
                                    []
                                    (Var 4 _lfortran_scos)
                                    Public
                                    .false.
                                    .false.
                                    ()
                                ),
                            _lfortran_scos__lfortran_scos:
                                (Variable
                                    3
                                    _lfortran_scos__lfortran_scos
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Real 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            _lfortran_scos__lfortran_scos1:
                                (Variable
                                    3
                                    _lfortran_scos__lfortran_scos1
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Real 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            x:
                                (Variable
                                    3
                                    x
                                    []
                                    In
                                    ()
                                    ()
                                    Default
                                    (Real 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            x__lfortran_scos:
                                (Variable
                                    3
                                    x__lfortran_scos
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Real 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            x__lfortran_scos1:
                                (Variable
                                    3
                                    x__lfortran_scos1
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Real 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    _lcompilers_cos_f32
                    (FunctionType
                        [(Real 4)]
                        (Real 4)
                        Source
                        Implementation
                        ()
                        .false.
                        .false.
                        .false.
                        .false.
                        .false.
                        []
                        .false.
                    )
                    [_lfortran_scos]
                    [(Var 3 x__lfortran_scos)]
                    [(=
                        (Var 3 _lcompilers_cos_f32)
                        (FunctionCall
                            3 _lfortran_scos
                            3 _lfortran_scos
                            [((Var 3 x__lfortran_scos))]
                            (Real 4)
                            ()
                            ()
                        )
                        ()
                    )]
                    (Var 3 _lcompilers_cos_f32)
                    Public
                    .false.
                    .false.
                    ()
                ),
            expr2:
                (Program
                    (SymbolTable
                        2
                        {
                            x:
                                (Variable
                                    2
                                    x
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Integer 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    expr2
                    []
                    [(=
                        (Var 2 x)
                        (IntegerConstant 3 (Integer 4))
                        ()
                    )
                    (Print
                        ()
                        [(FunctionCall
                            1 _lcompilers_cos_f32
                            1 _lcompilers_cos_f32
                            [((RealBinOp
                                (Cast
                                    (IntegerBinOp
                                        (Var 2 x)
                                        Add
                                        (IntegerConstant 4 (Integer 4))
                                        (Integer 4)
                                        ()
                                    )
                                    IntegerToReal
                                    (Real 4)
                                    ()
                                )
                                Mul
                                (RealConstant
                                    7.000000
                                    (Real 4)
                                )
                                (Real 4)
                                ()
                            ))]
                            (Real 4)
                            ()
                            ()
                        )]
                        ()
                        ()
                    )]
                )
        })
    []
)
```

</details>

**Review by certik — APPROVED** [review]
Looks good to me, if @czgdp1807 approves the fix.

**Review by czgdp1807 — APPROVED** [review]

**Pull request merged**